languages/emf-metamodel/src/main/java/de/jplag/emf/util/EmfaticModelView.java

```diff
@@ -0,0 +1,194 @@
+     */
+    private final List<String> generateEmfaticCode(StringBuilder builder, Resource modelResource) {
+        Writer writer = new Writer();
+        String code = writer.write(modelResource, null, null);
+        builder.append(code);
+        return builder.toString().lines().toList();
+    }
+
+    /**
+     * Calculates the line index of a metamodel token from the emfatic code. If it cannot be found, the last index is used.
+     */
+    private int calculateLineIndexOf(MetamodelToken token) {
+        int line = Token.NO_VALUE;
+        Optional<EObject> optionalEObject = token.getEObject();
+        if (optionalEObject.isPresent()) {
+            EObject eObject = optionalEObject.get();
+            if (eObject instanceof ENamedElement element) {
+                line = lineIndexOf(element);
+                if (line != Token.NO_VALUE && isEndToken(token)) {
+                    line = findEndIndexOf(line);
+                }
+            }
+        }
+        if (line == Token.NO_VALUE) {
+            return lastLineIndex;
+        }
+        lastLineIndex = line;
+        return line;
+    }
+
+    /**
+     * Locates the end index (closing character) for an element with a already known declaration index.
+     */
+    private int findEndIndexOf(int declarationIndex) {
+        int indentation = indentationOf(lines.get(declarationIndex));
+        if (declarationIndex > 1) { // exception for top level package
+            for (int i = declarationIndex + 1; i < lines.size(); i++) {
+                String nextLine = lines.get(i);
+                if (nextLine.length() > indentation && CLOSING_CHAR == nextLine.charAt(indentation)) {
+                    return i;
+                }
+            }
+        }
+        return lastLineIndex;
+    }
+
+    /**
+     * Checks if a token is representing a end of a block, e.g. a closing bracket.
+     */
+    private boolean isEndToken(Token token) {
+        return token.getType()instanceof MetamodelTokenType type && type.isEndToken();
+    }
+
+    /**
+     * Calculates the indentation of a line, meaning the number of tabs and spaces.
+     */
+    private int indentationOf(String beginLine) {
+        return beginLine.indexOf(beginLine.stripLeading());
+    }
+
+    /**
+     * Returns the line index of the declaration of an element in the Emfatic code.
+     */
+    private int lineIndexOf(ENamedElement element) {
+        return elementToLine.computeIfAbsent(element, this::findLineIndexOf);
+    }
+
+    /**
+     * Searches for the declaration of an element in the Emfatic code.
+     */
+    private int findLineIndexOf(ENamedElement element) {
+        String hash = Integer.toString(modelCopier.get(element).hashCode());
+        for (int index = 0; index < hashedLines.size(); index++) {
+            String line = hashedLines.get(index);
+            String trimmedLine = line.substring(indentationOf(line));
```

Comment:
I would do the indentation removal already when creating the `hashedLines`. As you already get a stream when calling `toLines()`, you can simply add `.map(l -> l.substring(indentationOf(l))` there.